hazelcast-client-new/src/main/java/com/hazelcast/client/txn/proxy/xa/XAResourceProxy.java

```diff
@@ -0,0 +1,239 @@
+                }
+                break;
+            case TMSUCCESS:
+            case TMSUSPEND:
+                break;
+            default:
+                throw new XAException("Unknown flag!!! " + flags);
+        }
+    }
+
+    @Override
+    public synchronized int prepare(Xid xid) throws XAException {
+        List<TransactionContext> contexts = xidContextMap.get(xid);
+        if (contexts == null) {
+            throw new XAException("No context with the given xid: " + xid);
+        }
+        for (TransactionContext context : contexts) {
+            XATransactionProxy transaction = getTransaction(context);
+            transaction.prepare();
+        }
+        return XA_OK;
+    }
+
+    @Override
+    public synchronized void commit(Xid xid, boolean onePhase) throws XAException {
+        List<TransactionContext> contexts = xidContextMap.remove(xid);
+        if (contexts == null && onePhase) {
+            throw new XAException("No context with the given xid: " + xid);
+        }
+        if (contexts == null) {
+            finalizeTransactionRemotely(xid, true);
+            return;
+        }
+
+        for (TransactionContext context : contexts) {
+            XATransactionProxy transaction = getTransaction(context);
+            transaction.commit(onePhase);
+        }
+    }
+
+    private void finalizeTransactionRemotely(Xid xid, boolean isCommit) {
+        SerializableXID serializableXID =
+                new SerializableXID(xid.getFormatId(), xid.getGlobalTransactionId(), xid.getBranchQualifier());
+        SerializationService serializationService = getContext().getSerializationService();
+        Data xidData = serializationService.toData(serializableXID);
+        ClientMessage request = XAFinalizeTransactionParameters.encode(xidData, isCommit);
+        invoke(request, xidData);
+    }
+
+    @Override
+    public synchronized void rollback(Xid xid) throws XAException {
+        List<TransactionContext> contexts = xidContextMap.remove(xid);
+        if (contexts == null) {
+            finalizeTransactionRemotely(xid, true);
+            return;
+        }
+        for (TransactionContext context : contexts) {
+            getTransaction(context).rollback();
+        }
+    }
+
+    @Override
+    public synchronized void forget(Xid xid) throws XAException {
+        throw new XAException(XAException.XAER_PROTO);
+    }
+
+    @Override
+    public synchronized boolean isSameRM(XAResource xaResource) throws XAException {
+        if (this == xaResource) {
+            return true;
+        }
+        String otherGroupName = null;
+        if (xaResource instanceof XAResourceProxy) {
+            otherGroupName = ((XAResourceProxy) xaResource).getGroupName();
+        }
+        if (xaResource instanceof XAResourceImpl) {
+            otherGroupName = ((XAResourceImpl) xaResource).getGroupName();
+        }
+        return getGroupName().equals(otherGroupName);
```

Comment:
As far as I remember, client doesn't have to use `group-name` as its login credentials.